### PR TITLE
add support file Xcode input file lists

### PR DESCRIPTION
**Pull request by BlueVirusX**
Fixes https://github.com/realm/SwiftLint/issues/5195

adds option "--use-script-input-file-lists"
adds parsing of .xcfilelist files

### CODE REVIEW

File: Source/SwiftLintFramework/Configuration+CommandLine.swift [hunk 1]
```diff
@@ -220,8 +256,8 @@ extension Configuration {
             }
             throw SwiftLintError.usageError(description: "stdin isn't a UTF8-encoded string")
         }
-        if options.useScriptInputFiles {
-            let files = try scriptInputFiles()
+        if options.useScriptInputFiles || options.useScriptInputFileLists {
+            let files = options.useScriptInputFiles ? try scriptInputFiles() : try scriptInputFileLists()
```
Comment (SimplyDanny) [inline Source/SwiftLintFramework/Configuration+CommandLine.swift:260]:
```suggestion
            let files = try options.useScriptInputFiles 
                ? scriptInputFiles()
                : scriptInputFileLists()
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -15,6 +15,10 @@
   types.  
   [Jordan Rose](https://github.com/jrose-signal)
   [SimplyDanny](https://github.com/SimplyDanny)
+* Support linting from Input File Lists provided by Run Script Phase of Xcode with
+  `--use-script-input-file-lists`.  
+  [BlueVirusX](https://github.com/BlueVirusX)
+  [#5790](https://github.com/realm/SwiftLint/pull/5790)
```
Comment (SimplyDanny) [inline CHANGELOG.md:21]: IDs refer to issues not to pull request. The PR ID will be attached to the commit message automatically upon merge.

```suggestion

* Support reading files to lint from Input File Lists provided
  by Run Script Build Phases in Xcode using the command-line
  argument `--use-script-input-file-lists`.  
  [BlueVirusX](https://github.com/BlueVirusX)
```

File: README.md [hunk 3]
```diff
@@ -513,7 +513,11 @@ plugin, or modified files in the working tree based on `git ls-files -m`), you
 can do so by passing the option `--use-script-input-files` and setting the
 following instance variables: `SCRIPT_INPUT_FILE_COUNT`
 and `SCRIPT_INPUT_FILE_0`, `SCRIPT_INPUT_FILE_1`, ...,
-`SCRIPT_INPUT_FILE_{SCRIPT_INPUT_FILE_COUNT - 1}`.
+`SCRIPT_INPUT_FILE_{SCRIPT_INPUT_FILE_COUNT - 1}` or by passing the
+option `--use-script-input-file-lists` and setting the
+following instance variables: `SCRIPT_INPUT_FILE_LIST_COUNT`
+and `SCRIPT_INPUT_FILE_LIST_0`, `SCRIPT_INPUT_FILE_LIST_1`, ...,
+`SCRIPT_INPUT_FILE_LIST_{SCRIPT_INPUT_FILE_LIST_COUNT - 1}`.
```
Comment (SimplyDanny) [inline README.md:520]:
```suggestion
`SCRIPT_INPUT_FILE_{SCRIPT_INPUT_FILE_COUNT - 1}`.
Similarly, files can be read from file lists by passing
the option `--use-script-input-file-lists` and setting the
following instance variables: `SCRIPT_INPUT_FILE_LIST_COUNT`
and `SCRIPT_INPUT_FILE_LIST_0`, `SCRIPT_INPUT_FILE_LIST_1`, ...,
`SCRIPT_INPUT_FILE_LIST_{SCRIPT_INPUT_FILE_LIST_COUNT - 1}`.
```

File: Source/swiftlint/Common/LintOrAnalyzeArguments.swift [hunk 4]
```diff
@@ -31,6 +31,8 @@ struct LintOrAnalyzeArguments: ParsableArguments {
     var useAlternativeExcluding = false
     @Flag(help: "Read SCRIPT_INPUT_FILE* environment variables as files.")
     var useScriptInputFiles = false
+    @Flag(help: "Read SCRIPT_INPUT_FILE_LIST* environment variables as file list.")
```
Comment (SimplyDanny) [inline Source/swiftlint/Common/LintOrAnalyzeArguments.swift:34]:
```suggestion
    @Flag(help: "Read SCRIPT_INPUT_FILE_LIST* environment variables as file lists.")
```

File: Source/SwiftLintFramework/Configuration+CommandLine.swift [hunk 5]
```diff
@@ -39,6 +39,42 @@ private func scriptInputFiles() throws -> [SwiftLintFile] {
     }
 }
 
+private func scriptInputFileLists() throws -> [SwiftLintFile] {
```
Comment (SimplyDanny) [inline Source/SwiftLintFramework/Configuration+CommandLine.swift:42]:
```suggestion
private func readFilesFromScriptInputFileLists() throws -> [SwiftLintFile] {
```

Please rename the existing method accordingly.

File: Source/SwiftLintFramework/Configuration+CommandLine.swift [hunk 6]
```diff
@@ -39,6 +39,42 @@ private func scriptInputFiles() throws -> [SwiftLintFile] {
...
+    let inputFileListsKey = "SCRIPT_INPUT_FILE_LIST_COUNT"
+    guard let countString = ProcessInfo.processInfo.environment[inputFileListsKey] else {
+        throw SwiftLintError.usageError(description: "\(inputFileListsKey) variable not set")
+    }
+
+    guard let count = Int(countString) else {
+        throw SwiftLintError.usageError(description: "\(inputFileListsKey) did not specify a number")
+    }
```
Comment (SimplyDanny) [inline Source/SwiftLintFramework/Configuration+CommandLine.swift:50]: Extract this to a method accepting a variable name and retuning an integer.

File: Source/SwiftLintFramework/Configuration+CommandLine.swift [hunk 7]
```diff
@@ -39,6 +39,42 @@ private func scriptInputFiles() throws -> [SwiftLintFile] {
...
+
+    return (0..<count).flatMap { fileNumber in
+        var swiftLintFiles: [SwiftLintFile] = []
```
Comment (SimplyDanny) [inline Source/SwiftLintFramework/Configuration+CommandLine.swift:53]:
```suggestion
        var filesToLint: [SwiftLintFile] = []
```

**Review by SimplyDanny — CHANGES_REQUESTED** [review]
Seems reasonable to add this. Thanks for the implementation!

I've added a few remarks. Please take them into account.

File: Source/SwiftLintFramework/Configuration+CommandLine.swift [hunk 8]
```diff
@@ -39,6 +31,44 @@ private func scriptInputFiles() throws -> [SwiftLintFile] {
...
+                }
+            }
+        } catch {
+            queuedPrintError(String(describing: error))
+        }
+        return filesToLint
+    }
+}
+
+private func fileCount(for envVar: String) throws -> Int {
```
Comment (SimplyDanny) [inline Source/SwiftLintFramework/Configuration+CommandLine.swift:62]:
```suggestion
private func fileCount(from envVar: String) throws -> Int {
```

**Comment by BlueVirusX** [comment]
Thanks for the review @SimplyDanny!

**Review by SimplyDanny — APPROVED** [review]
Perfect! Thanks for the updates.

**Pull request merged**